File: ProductExtractor/models/offer.py
from models.builder import MissingAttributeException
from models.category import Category
from models.product import Product, ProductBuilder
from models.provider import Provider


class Offer:
    def __init__(
        self,
        product: Product,
        provider: Provider,
        price: float,
        link: str,
    ):
        self.product = product
        self.provider = provider
        self.price = price
        self.link = link

    def __repr__(self) -> str:
        res = "ProductOffer(\n"
        for key, val in self.__dict__.items():
            res += f"\t{key} = {str(val)}\n"
        res += ")"
        return res


class OfferBuilder:
    def __init__(self, product: ProductBuilder, price: float, url: str) -> None:
        self.product = product
        self.provider = None
        self.price = price
        self.url = url

    def set_provider(self, provider: Provider):
        self.provider = provider

    def build(self) -> Offer:
        if self.provider is None:
            raise MissingAttributeException(
                "Tried to contruct Offer without setting provider attribute"
            )

        product = self.product.build()

        return Offer(product, self.provider, self.price, self.url)

File: offer-extractor/models/offer.ts
import { ProductBuilder, Product } from "./product.ts";
import { Provider } from "./provider.ts";


export class Offer {
    product: Product;
    provider: Provider;
    price: number;
    link: string;


    constructor(
        product: Product,
        provider: Provider,
        price: number,
        link: string,
    ) {
        this.product = product
        this.provider = provider
        this.price = price
        this.link = link
    }

    toString(): string {
        let res = "Offer(\n"
        Object.entries(this).forEach((key, val) => {
            res += `\t${key} = ${val}\n`;
        })
        res += ")"
        return res
    }
}

export class OfferBuilder {
    product: ProductBuilder;
    price: number;
    url: string;
    provider?: Provider;

    constructor(product: ProductBuilder, price: number, url: string) {
        this.product = product;
        this.price = price
        this.url = url
    }

    setProvider(provider: Provider) {
        this.provider = provider
    }

    build(): Offer {
        if (this.provider == null) {
            throw Error(
                "Tried to contruct Offer without setting provider attribute"
            )
        }


        const product = this.product.build()

        return new Offer(product, this.provider, this.price, this.url)
    }
}
